---
title: "Database"
author: "Abelardo"
date: "2023-01-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Dependencies
library(tidyverse)
library(taxize)

# Variables
## Input
anytaxon.csv <- "data/refseq_circoviruses.csv" #path to initial file in csv format containing information of a given taxon
taxa_column <- "Species"
## Include host ?
include_hosts <- TRUE
host_taxa_column <- "Host"
## Output
export_db <- TRUE
output.csv <- "results/refseq_circoviruses_expanded_clasif.csv" #path to initial file in csv format containing information of a given taxon

```

#### Recursive pangenomics

... is a revision of the original ssDNA viral pangenomics project. In "Recursive pangenomics I'll try to revisit and rethink the original pipeline that departed from a set of genomes and was able to fulfill the tasks of protein count filtering, segment concatenation, pre-pangenomic clustering, pangenomic analysis, functional annotation, remote homologues search and phylogenetic inference of nuclear pangenomic elements.

Several steps from the original pipeline were reviewed, some will be added while others were eliminated. In particular, to the date of january 2023 those of protein count filtering and functional annotation were completely removed. Moreover, the pre-pangenomic clustering was replaced by a whole new scope, a recursive pangenomics guided from the tips of a reference tree to its roots(?)... if such things exist.

As a general description, these pangenomic analysis will start from the lowest hierarchy level of a given phylogeny (lets call it h1) and advance in a staged process where the originals genomes of h1 are compared in a first pangenomic round and then replaced by the resulting pangenomes for the hierarchical level 2 (h2) and so on until the hypothetical root is reached. To determine which proteins will represent its cluster in the following hierarchy level, recursive medioid determinations will be performed.

## Database 

As the original project, recursive pangenomics will be focused on the evolutionary exploration of viruses with the goal of finding clusters of homologus proteins and their prevalence within lineages.

We will begin by properly reading and formatting a table of ssDNA viruses into a tibble.

```{r anytaxon.tb, error=FALSE, warning=FALSE, message=FALSE}
anytaxon.tb <- readr::read_csv(anytaxon.csv)
```

One first goal is to enrich our tibble with taxonomic information for both. Such goal will be reached with the local function get_expanded_classification, which is a function that depends on {taxize} and {tidyverse}. We will import it from source.

```{r source_functions}
source("Functions.R")
```

It takes a data frame and a column name as input and then retrieves the NCBI taxonomy of all the elements in the selected column and returns and expanded data frame where each row correspond to a unique query, and each column to its assigned taxonomy at each rank.

For the moment we will only work with ssDNA viruses, to stay with low computation resources. So we make a search from our database:

The following chunk invokes get_expanded_classification() to retrieve the full taxonomic classification of anytaxon.csv members and it also merges the outcome with the initial tibble. 

```{r retrieve_classification, error=FALSE, warning=FALSE, message=FALSE}
anytaxon_classification.df <- get_expanded_classification(anytaxon.tb, taxa_column)
merge_anytaxon.df <- base::merge(anytaxon.tb, anytaxon_classification.df, by.x = taxa_column, by.y = taxa_column, no.dups = FALSE)
```

As we will be mainly working with viruses, a optional chunk was added to retrieve and merge the taxonomic classification of hosts too.

```{r if_include_hosts, error=FALSE, warning=FALSE, message=FALSE}
if(include_hosts){
  hosts_classification.df <- get_expanded_classification(anytaxon.tb, host_taxa_column)
  merge_anytaxon.df <- base::merge(merge_anytaxon.df, hosts_classification.df, by.x = host_taxa_column, by.y = host_taxa_column, no.dups = FALSE)
}
```

At this point we have generated an extensive database enriched with classifications according to NCBI taxonomy, which we can export running the following chunk.

```{r database_export, error=FALSE, warning=FALSE, message=FALSE}
if(export_db){write_csv(merge_anytaxon.df, file = output.csv)}
```






